src/timeutils.cc

```diff
@@ -0,0 +1,184 @@
+#include "winutils.h"
+#include <winsock2.h> //for timeval
+#include<time.h>
+
+#include <iomanip>
+#include <sstream>
+#include <sys/time.h>
+#include <cassert>
+
+#define BUF_SIZE 128
+//C++ 11: Lot of OS specific functions can be replaced by using the boost::chrono library
+
+bool ValidClockId(clockid_t clk_id) {
+    if (clk_id == CLOCK_MONOTONIC || clk_id == CLOCK_REALTIME)
+        return true;
+    else
+        return false;
+}
+
+int clock_getres_monotonic( struct timespec *res) {
+    if (res != nullptr) {
+        LARGE_INTEGER frequency;
+        if (QueryPerformanceFrequency(&frequency) != 0) {
+            double nanosecdbl = 1000000000.0 / frequency.QuadPart;
+            res->tv_nsec = static_cast<long>(nanosecdbl) % 1000000000;
+            res->tv_sec = static_cast<long>(nanosecdbl) / 1000000000;
+            return 0;
+        }
+    }
+    return -1;
+}
+
+int clock_getres(clockid_t clk_id, struct timespec *res) {
+    if (clk_id == CLOCK_MONOTONIC)
+        return clock_getres_monotonic(res);
+    return -1;
+}
+
+int clock_gettime_monotonic(struct timespec *ts) {
+    static bool bInitDone = false;
+    if (ts == nullptr)
+        return -1;
+
+    LARGE_INTEGER           t;
+//	FILETIME            f;
+    double                  nanoseconds;
+    static LARGE_INTEGER    starttime;
+    static double           TicksPerNanosecond;
+
+    if (!bInitDone) {
+        QueryPerformanceCounter(&starttime);
+
+        LARGE_INTEGER performanceFrequency;
+        QueryPerformanceFrequency(&performanceFrequency);
+        bInitDone = true;
+        TicksPerNanosecond = (double)performanceFrequency.QuadPart / 1000000000.;
+
+    }
+
+    QueryPerformanceCounter(&t);
+    t.QuadPart -= starttime.QuadPart;
+
+    nanoseconds = (double)t.QuadPart / TicksPerNanosecond;
+    ts->tv_nsec = static_cast<long>(nanoseconds) % 1000000000;
```

Comment:
I think modulo'ing/dividing first and then converting to long would be a better approach. Right now the maximum supported time equal to about 2.1 seconds.